Apromore-Custom-Plugins/Process-Discoverer-Portal-Plugin/src/main/java/org/apromore/plugin/portal/processdiscoverer/impl/apmlog/LogFilterControllerWithAPMLog.java

```diff
@@ -24,55 +24,193 @@
-                        parent.getLogData().getLog(), 
-                        parent.getUserOptions().getMainAttributeKey(), 
-                        parent.getLogData().getCurrentFilterCriteria()),
-                        this);
-
-        subscribeFilterResult();
+        if (event.getData() == null) {
+            LogFilterRequest lfr = logData.getCurrentFilterCriteria() == null ||
+                    ((List<LogFilterRule>) logData.getCurrentFilterCriteria()).isEmpty() ?
+                    getRequestWithOption(new EditorOption(FilterType.CASE_VARIANT)) : getDefaultRequest();
+            parent.getLogFilterPlugin().execute(lfr);
+        } else {
+            if (event.getData() instanceof JSONObject) onInvokeExtEvent((JSONObject) event.getData());
+            else onInvokeEvent(event.getData().toString());
+        }
+    }
+
+    private void onInvokeEvent(String payload) {
+        FilterType filterType = getFilterType(payload);
+        LogFilterRule rule = getLastMatchedRule(filterType, (List<LogFilterRule>) logData.getCurrentFilterCriteria());
+        EditorOption option = rule != null ? new EditorOption(filterType, rule) : new EditorOption(filterType);
+        LogFilterRequest lfr = getRequestWithOption(option);
+        parent.getLogFilterPlugin().execute(lfr);
+    }
+
+    private void onInvokeExtEvent(JSONObject param) {
+        String type = (String) param.get("type");
+        FilterType filterType = getFilterType(type);
+        String data, source, target;
+        Map<String, Object> parameters = new UnifiedMap<>();
+        String mainAttribute = parent.getUserOptions().getMainAttributeKey();
+        if (filterType == FilterType.CASE_SECTION_ATTRIBUTE_COMBINATION ||
+                filterType == FilterType.EVENT_ATTRIBUTE_DURATION) {
+            data = (String) param.get("data");
+            if (filterType == FilterType.EVENT_ATTRIBUTE_DURATION &&
+                    !logData.hasSufficientDurationVariant(mainAttribute, data)) {
+                Messagebox.show("Unable to filter on node duration as there's only one value.",
+                        "Filter error", Messagebox.OK, Messagebox.ERROR);
+                return;
+            }
+            parameters.put("key", mainAttribute);
+            parameters.put("value", data);
+        } else if (filterType == FilterType.ATTRIBUTE_ARC_DURATION) {
+            source = (String) param.get("source");
+            target = (String) param.get("target");
+            if (!logData.hasSufficientDurationVariant(mainAttribute, source, target)) {
+                Messagebox.show("Unable to filter on arc duration as there's only one value.",
+                        "Filter error", Messagebox.OK, Messagebox.ERROR);
+                return;
+            }
+            parameters.put("key", mainAttribute);
+            parameters.put("from", source);
+            parameters.put("to", target);
+        } else {
+            return;
+        }
```

Comment:
This may be a little bit clearer as a switch statement, e.g.
```
switch (filterType) {
case CASE_SECTION_ATTRIBUTE_COMBINATION:
case EVENT_ATTRIBUTE_DURATION:
   ....
   break;
case ATTRIBUTE_ARC_DURATION:
    ....
    break;
default:
    return;
}